### PR TITLE
AI Cyborg-Shells

### PR DESCRIPTION
**Pull request by gurfan**
This is a feature that TG has that I wanted to try implementing here. I ported over some of the code from TG but with a few mechanical changes.

AIs can now swap between their core and a cyborg shell! 
AIs now have an action button in the top right corner of their screen that allows them to swap to a cyborg body.
The button does one of two things when clicked:

- If the AI has an active (living) cyborg body, then they will possess it.
- If the AI doesn't have a cyborg body, then it will look for any completed robot exoskeletons that are visible from cameras. A list will pop up containing each exoskeleton and stating its location. Selecting an exoskeleton from this list will make that your new shell.

The AI can only have one cyborg body at a time.  In order to get a new one, the old body must be destroyed, which the AI can do at any time using an action button.

Cyborg Shells are functionally identical to normal cyborgs, but with a few noticeable changes:

- Whenever cyborg shells are emagged, they gib instead of giving the AI a syndicate lawset.
- Having your AI Link wires cut will lock down the shell. The AI can still detonate the shell in order to activate a new one. 

This PR also adds in a completed-exoskeleton object that can be used to map or spawn in completed exoskeletons.
Exoskeletons can now be smashed by any weapon dealing at least 15 force (toolbox) or by bullets/lasers dealing at least 10 damage. Destroying the exoskeleton will destroy the frame but drop all of the added parts (arms, chest, legs, ect) on the ground.

Picture Gallery: 

Pete Hangar XVII 2020-08-31 173221

To-Dos:

- [x]  Allow AIs to click on exoskeletons to create shells, rather than always having to use the menu.

- [x]  ~~Add one complete exoskeleton to the AI cores.~~ Instead of always having a shell ready for the AI, one will spawn if no cyborgs readied up.

- [x]  Allow exoskeletons to be smashed apart.

- [x] Maybe have the action button display whatever customized AI core display you are using, instead of always the default one.

- [x] Custom job title for shells.

- [x] Make sure losing your core will kill you (if you are in a shell)

- [x] Make sure intellicards dont break everything.

- [x] ~~Green~~ White eye color for shells

- [x] Test if malf AIs can use spells while in a shell and make sure nothing breaks, especially when they shunt.

[content]

:cl:
 * rscadd: AI's can now swap between a cyborg shell and their main core!
 * rscadd: AI's can now create cyborg shells from completed cyborg skeletons that are visible to the AI. The AI can only have one shell at a time; the previous one must be destroyed before a new one can be made. The AI can gib their shell at any time using an action button. (Even while the shell is dead!)
 * rscadd: AI-Cyborg Shells have a unique title on the radio and a white glow to their eyes, unlike normal cyborgs. When emagged, they will gib instead of receiving a syndicate lawset.
 * rscadd: Completed cyborg exoskeletons can be touched with an empty hand to prevent/allow the AI to turn it into a shell.
 * rscadd: Cyborg exoskeletons can now be smashed apart with anything as strong as a toolbox. Doing so will destroy the exoskeleton but drop any attached parts (arms, legs, torso, head) on the ground. It will also make a good mess for the janitor to clean up.

### CODE REVIEW

**Comment by mrflarg** [comment]
what job title does the cyborg get on the radio?

**Comment by gurfan** [comment]
Their job says cyborg currently, I will work on getting them a custom title

**Comment by Blithering** [comment]
Does the AI need to destroy its body to return to the core?

**Comment by cloutcode** [comment]
1man malf challenge

**Comment by gurfan** [comment]
> Does the AI need to destroy its body to return to the core?

No, they can swap between the cyborg and the core at any time. The only time they need to destroy the cyborg is to create a new one.

**Comment by SonixApache** [comment]
OP as fuck but i don't even care, i want it

**Comment by DrSnips** [comment]
Why do they gib on death rather than become a borg "husk"?

**Comment by gurfan** [comment]
I wanted some downsides to shells compared to regular borgs, and I also wanted AIs to immediately have the ability to create a second shell. It might not be that important since I added the ability for the AI to destory its shell at any time, so I can revert it to normal. Now that I look back at it the next day I don't really know remember why I made it that way either.

**Comment by tuypo1** [comment]
Is there any visual difference? because emaging what you think is a borg only to suddenly blow your cover when it turns out to have been the AI sounds super lame

**Comment by gurfan** [comment]
> Is there any visual difference? because emaging what you think is a borg only to suddenly blow your cover when it turns out to have been the AI sounds super lame

Right now their name is always Shell of [AI name]. I will also give them a separate job title and recolor their "eye glow" to make them more distinguishable.

**Comment by toomykins** [comment]
Both cool and OP as fuck, I'd thought of giving the AI more control over dumb little things like farmbots to give it more control over the station. This could do the trick, maybe it'd be better if they were a little more limited.
Also recommend changing [name] Shell to Shell of [name], for readability.

**Comment by optimumtact** [comment]
good feature :+1:

**Comment by cloutcode** [comment]
bad feature

**Comment by toomykins** [comment]
dunno if you already have, but also recommend testing this with the staff of change, since mind.transfer_to can be fucky
Edit; Did a quick test myself. the staff of change does work on the borg, completely detaching you from your AI core and turning you into the resulting mob.

**Comment by tuypo1** [comment]
> > Is there any visual difference? because emaging what you think is a borg only to suddenly blow your cover when it turns out to have been the AI sounds super lame
> 
> Right now their name is always Shell of [AI name]. I will also give them a separate job title and recolor their "eye glow" to make them more distinguishable.

Sounds good then

**Comment by gurfan** [comment]
> dunno if you already have, but also recommend testing this with the staff of change, since mind.transfer_to can be fucky
> Edit; Did a quick test myself. the staff of change does work on the borg, completely detaching you from your AI core and turning you into the resulting mob.

fug, thanks for pointing that out, I'll fix it

**Comment by Kammerjunk** [comment]
hahaha holy shit
if this won't give credence to people complaining about the AI role, I don't know what will

**Comment by SonixApache** [comment]
~~Just to check - AIs can't just produce shells for themselves, right? You need the help of a carbon, right? Because if not, you need to unfuck that.~~ Okay, I read the OP. Sorry.
Maybe it'd be better if carbons had to install a remote-control device in place of the usual MMI for the AI to be able to take over the exoskeleton.

As for the "add a skeleton to each map", I recommend just using the job slot system for that. Make it so there's always two "borgs" for the AI - if there's no player available for the borg, spawn an AI-controllable one.
Would help you avoid map conflicts and other mapping horror stories.

**Comment by gurfan** [comment]
> Maybe it'd be better if carbons had to install a remote-control device in place of the usual MMI for the AI to be able to take over the exoskeleton.

Not a bad idea, one of the things I can see happening is that robotics will stop leaving pre-built cyborgs sitting there in case the AI snatches them. I'm not exactly sure what the last device would be. Something like a tracking beacon might work well but then the roboticists will have to know that they need to run to R&D and jam one in the exoskeleton. Clicking on the exoskeleton with an empty hand to disable/enable AI control might be a more intuitive option, and it wouldn't force people to smash their exoskeletons to bits every time the AI is rogue either.

> As for the "add a skeleton to each map", I recommend just using the job slot system for that. Make it so there's always two "borgs" for the AI - if there's no player available for the borg, spawn an AI-controllable one.
> Would help you avoid map conflicts and other mapping horror stories.

This is a great idea and I'm already having mapping nightmares. One of the things I was worried about was that AI's would ignore ordering around their borgs and watching over the station like a proper AI in favor of just playing as a more powerful cyborg. The purpose of this PR was mainly to ease some of the boredom/inability to act that comes when you're playing as a borgless AI.

**Comment by gurfan** [comment]
All the shells will have a white eye-glow. Since each custom cyborg look has a different eye color, white is really the only color that will always stand out and let people know that its a shell. If anyone has better ideas let me know.

File: code/modules/power/apc.dm [hunk 1]
```diff
@@ -941,7 +941,10 @@
 						src.malfai = usr:parent
 					else
 						src.malfai = usr
-					to_chat(malfai, "Hack complete. The APC is now under your exclusive control. [STATION_Z == z?"You now have [M.apcs] under your control.":"As this APC is not located on the station, it is not contributing to your control of it."]")
+					if(malfai.deployed)
+						to_chat(malfai.shell, "Hack complete. The APC is now under your exclusive control. [STATION_Z == z?"You now have [M.apcs] under your control.":"As this APC is not located on the station, it is not contributing to your control of it."]")
+					else
+						to_chat(malfai, "Hack complete. The APC is now under your exclusive control. [STATION_Z == z?"You now have [M.apcs] under your control.":"As this APC is not located on the station, it is not contributing to your control of it."]")
```
Comment (jknpj) [inline code/modules/power/apc.dm]:
```suggestion
					var/target_malf = malfai.shell || malfai
						to_chat(target_malf, "Hack complete. The APC is now under your exclusive control. [STATION_Z == z?"You now have [M.apcs] under your control.":"As this APC is not located on the station, it is not contributing to your control of it."]")
						
```
maybe something like this?

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 2]
```diff
@@ -1006,8 +1010,12 @@ var/list/cyborg_list = list()
 	overlays.Cut()
 	update_fire()
 	if(!stat && cell != null)
-		eyes = image(icon,"eyes-[icon_state]", overlay_layer)
+		var/icon/eyesicon = icon(icon,"eyes-[icon_state]", overlay_layer)
+		if(isshell(src))	//Make them white.
```
Comment (jknpj) [inline code/modules/mob/living/silicon/robot/robot.dm]: I fucking hate src checks. It's MoMMI hell all over again

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 3]
```diff
@@ -557,13 +557,17 @@ var/list/cyborg_list = list()
 				to_chat(user, "The wires get in your way.")
 			else
 				if(prob(50))
+					to_chat(user, "You emag [src]'s interface")
+					message_admins("[key_name_admin(user)] emagged cyborg [key_name_admin(src)].")
+					if(isshell(src))	//Putting this here would be cleaner than shoving it in a child proc
```
Comment (jknpj) [inline code/modules/mob/living/silicon/robot/robot.dm]: No, just move this to the shell's SetEmagged or something.

File: code/modules/mob/living/silicon/robot/login.dm [hunk 4]
```diff
@@ -1,13 +1,15 @@
 /mob/living/silicon/robot/Login()
 	..()
 	regenerate_icons()
-	show_laws(0)
+	if(!isshell(src))
```
Comment (jknpj) [inline code/modules/mob/living/silicon/robot/login.dm]: Maybe do this in the shell's type? I don't even care if you have to copypaste, it's better than having src checks.

File: code/modules/mob/living/silicon/robot/laws.dm [hunk 5]
```diff
@@ -69,7 +69,11 @@
 				laws.supplied[index] = temp
 
 		if(mind)
-			var/datum/role/mastermalf = connected_ai.mind.GetRole(MALF)
+			var/datum/role/mastermalf		//Workaround for shelled AIs causing runtimes
+			if(connected_ai.deployed)
+				mastermalf =  connected_ai.shell.mind.GetRole(MALF)
+			else
+				mastermalf = connected_ai.mind.GetRole(MALF)
```
Comment (jknpj) [inline code/modules/mob/living/silicon/robot/laws.dm]:
```suggestion
				mastermalf =  connected_ai.shell.mind.GetRole(MALF) || connected_ai.mind.GetRole(MALF)
```
Maybe something like this?

File: code/modules/mob/living/silicon/robot/death.dm [hunk 6]
```diff
@@ -34,6 +34,10 @@
 	dead_mob_list -= src
 	qdel(src)
 
+/*
+/mob/living/silicon/robot/shell/death(gibbed)
+	..(TRUE)	//Always gib.
+*/
```
Comment (jknpj) [inline code/modules/mob/living/silicon/robot/death.dm]:
```suggestion
```
delet

**Comment by jknpj** [comment]
I got out of breath midway through the code, will take another look later.

**Comment by PrimeDSS13** [comment]
quick close it 
image

**Comment by gurfan** [comment]
conflicts fixed

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 7]
```diff
@@ -534,6 +534,7 @@
 	if (sterile)
 		to_chat(user, "<span class='danger'>It looks like \the [src]'s has been de-beaked.</span>")
 	return
+
```
Comment (jknpj) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm:537]: ?

**Comment by SonixApache** [comment]
:tada:

**Pull request merged**